Domain Display Name

yaleman opened the pull request:
- implements domain_display_name - a human-facing name for the system
- turned a repeated string path in UNIX tools to a constant
- reformatted design files reflexively while reading them

Fixes #860

- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes

yaleman commented on kanidm_unix_int/build.rs:
```diff
@@ -85,7 +85,7 @@ fn main() {
         &mut UnixdStatusOpt::command(),
         "kanidm_unixd_status",
-        comp_dir.clone(),
+        comp_dir,
```
clippy smiles on thee, tiny change.

Firstyear commented on examples/insecure_server.toml:
```diff
@@ -11,5 +11,7 @@ tls_key = "/tmp/kanidm/key.pem"
 log_level = "verbose"
 
 domain = "idm.example.com"
+domain_display_name = "Kani's Crab Shack"
```
Not sure about this being in the server.toml, since this is a "domain wide" setting, not a "per-server" setting, and it's not critical to have in place for bootstrap (like the domain name is).

Firstyear commented on kanidm_proto/src/v1.rs:
```diff
@@ -776,6 +776,7 @@ impl fmt::Display for TotpAlgo {
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct TotpSecret {
     pub accountname: String,
+    /// TODO: #860 this'll be domain_display_name
```
It still needs to be called issuer though ....

yaleman replied:
Yep, this is just me dropping flags around the code going "hey make changes here"

Firstyear commented on kanidmd/daemon/src/main.rs:
```diff
@@ -55,6 +55,7 @@ struct ServerConfig {
     pub log_level: Option<String>,
     pub online_backup: Option<OnlineBackup>,
     pub domain: String,
+    pub domain_display_name: Option<String>,
```
Again, probably shouldn't be a setting here ...

Firstyear commented on kanidmd/idm/src/idm/credupdatesession.rs:
```diff
@@ -124,7 +126,9 @@ impl fmt::Debug for MfaRegStateStatus {
 pub struct CredentialUpdateSessionStatus {
+    // #860 can we shove it in here?
     spn: String,
+    // TODO: is this the user's display name, or the session, or some random crab that was running down the street?
```
User's displayname.

Firstyear commented on kanidmd/idm/src/idm/credupdatesession.rs:
```diff
@@ -295,6 +300,8 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
         // - store account policy (if present)
         let session = CredentialUpdateSession {
+            // #860 - here?
+            // domain_display_name = self.qs_write.get_domain_display_name()
```
👍

Firstyear commented on kanidmd/idm/src/idm/credupdatesession.rs:
```diff
@@ -413,6 +420,7 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
     pub fn exchange_intent_credential_update(
         &mut self,
         token: CredentialUpdateIntentToken,
+        //TODO: rename ct to something... else?
```
current_time is what ct stands for.

Firstyear commented on kanidmd/idm/src/idm/credupdatesession.rs:
```diff
@@ -587,16 +595,16 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
-
             // ==== AUTHORISATION CHECKED ===
-
+            // #860 we have access to the qs here
```
We can pass in the values we need to this function preferrabley over accessing the qs.

Firstyear commented on kanidmd/idm/src/idm/mfareg.rs:
```diff
@@ -107,20 +108,21 @@ impl MfaRegSession {
 
                     if token_sha1.verify(chal, ct) {
-                        // Greeeaaaaaatttt it's a broken app. Let's check the user
+                        // Greeeaaaaaatttt. it's a broken app. Let's check the user
```
Missed the capital 'I' on 'It's' if you're adding a '.' 🤣

Firstyear commented on kanidmd/score/src/https/v1.rs:
```diff
@@ -887,7 +887,8 @@ pub async fn auth(mut req: tide::Request<AppState>) -> tide::Result {
 
     let maybe_sessionid = req.get_current_auth_session_id();
-    debug!("🍿 {:?}", maybe_sessionid);
+    // TODO: lol, what is this
```
I don't remember?

Firstyear reviewed (changes requested):
So I think you should check out https://github.com/kanidm/kanidm/blob/master/kanidmd/idm/src/plugins/domain.rs . This is where if the "domain display name" is unset on the domain entry, then you can generate it with a default value here. That way it's *not* in the server.toml, and it's generated properly at startup/bootstrap/updates. 

Then in the get_domain_display_name, you'll access the "cached" version. See how domain_name gets reloaded and cached in the end of some operations. 

But yeah, I don't think it shuold come from a config file, for the risk of "fighting" between replicas / servers, and the potential for inconsistency between servers.

yaleman commented:
> So I think you should check out https://github.com/kanidm/kanidm/blob/master/kanidmd/idm/src/plugins/domain.rs . This is where if the "domain display name" is unset on the domain entry, then you can generate it with a default value here. That way it's _not_ in the server.toml, and it's generated properly at startup/bootstrap/updates.
> 
> Then in the get_domain_display_name, you'll access the "cached" version. See how domain_name gets reloaded and cached in the end of some operations.
> 
> But yeah, I don't think it shuold come from a config file, for the risk of "fighting" between replicas / servers, and the potential for inconsistency between servers.

Thanks, I'll go that way - I didn't consider the domain-wide replication bits. It will probably make it easier to access, as well 😄

Firstyear commented:
> Thanks, I'll go that way - I didn't consider the domain-wide replication bits. It will probably make it easier to access, as well 😄

Yep, and easier to admin too, since you can update in a single place :)

Firstyear commented on insecure_generate_tls.sh:
```diff
@@ -1,7 +1,14 @@
 
-KANI_TMP=/tmp/kanidm/
+# also where the files are stored
+if [ -z "$KANI_TMP" ]; then
```
At what point do you convert this script to python .....

yaleman replied:
sooooon... or a rust build target that makes and deploys the certs 🤔

Firstyear commented on insecure_generate_tls.sh:
```diff
@@ -95,4 +104,5 @@ openssl dhparam -in "${CAFILE}" -out "${DHFILE}" 2048
 echo "Private key is at: ${KEYFILE}"
-
+echo ""
+echo "**Remember** the default action is to store the files in /tmp/ so they'll be deleted on reboot! Set the KANI_TMP environment variable before running this script if you want to change that. You'll need to update server config elsewhere if you do, however."
```
Believe it or not on macos, if it hibernates it'll also clear this folder.

yaleman replied:
That's what you get for leaving your machine alone for so long. 😭

Firstyear commented on kanidmd/idm/src/be/mod.rs:
```diff
@@ -1456,6 +1456,7 @@ impl<'a> BackendWriteTransaction<'a> {
         Ok(nsid)
     }
 
+    /// This... gets the domain uuid?
```
Yes.

Firstyear commented on kanidmd/idm/src/constants/schema.rs:
```diff
@@ -214,6 +214,38 @@ pub const JSON_SCHEMA_ATTR_DOMAIN_NAME: &str = r#"{
+    ],
+    "index": [
+      "EQUALITY",
+      "PRESENCE"
```
I'm not sure you need presence here.

yaleman replied:
I may just be yeeting random things around while I figure out how to modify schema 🙄

Firstyear commented on kanidmd/idm/src/server.rs:
```diff
@@ -929,10 +937,17 @@ impl<'a> QueryServerTransaction<'a> for QueryServerWriteTransaction<'a> {
 
+    fn get_domain_display_name(&self) -> String {
+        // TODO: lol this is probably terrible
+        String::from("filler text from server.rs qst qstwt")
```
That's fine, you just need the cowcell later that does the reload on change thingo.

yaleman replied:
🐄 🤠

Firstyear commented:
Progress looks great! Sorry that the code is a bit all over the place :)

yaleman commented:
> Progress looks great! Sorry that the code is a bit all over the place :)

It's definitely a "if you know, you know" kind of internal thing, I dug into the designs folder and will pull some things out of there to merge into the developer section of the book so they're easier to consume for future-humans.

Firstyear commented:
> > Progress looks great! Sorry that the code is a bit all over the place :)
> 
> It's definitely a "if you know, you know" kind of internal thing, I dug into the designs folder and will pull some things out of there to merge into the developer section of the book so they're easier to consume for future-humans.

Well, I've also been on a bit of a "refactor" before the last release, and I'll do more soon. I want to kinda create a bit more order in the code base :)

Firstyear commented on kanidmd/daemon/src/opt.rs:
```diff
@@ -102,33 +142,29 @@ enum KanidmdOpt {
-    Verify(CommonOpt),
     #[clap(name = "recover_account")]
-    /// Recover an account's password
+    /// Recover an account's password (must be run offline)
```
Technically all of these have to .....

yaleman replied:
Weeeeelllllll the db things don't, but sure, removed. 😄

Firstyear commented on kanidmd/idm/src/actors/v1_read.rs:
```diff
@@ -44,7 +44,7 @@ use std::convert::TryFrom;
 pub struct QueryServerReadV1 {
     pub log_level: Option<u32>,
-    idms: Arc<IdmServer>,
+    pub idms: Arc<IdmServer>,
```
Why make this pub?

yaleman replied:
This was part of me chasing down weird paths, don't need it now - removed. 😄

Firstyear commented on kanidmd/idm/src/be/mod.rs:
```diff
@@ -1450,24 +1450,42 @@ impl<'a> BackendWriteTransaction<'a> {
             .get_idlayer()
             .get_db_d_uuid()
-            .expect("DBLayer Error!!!")
+            .expect("DBLayer Error pulling Domain UUID!!!")
```
retrieving?

Firstyear commented on kanidmd/idm/src/be/mod.rs:
```diff
@@ -1450,24 +1450,42 @@ impl<'a> BackendWriteTransaction<'a> {
+        //     None => self.reset_db_d_uuid().expect("Failed to regenerate D_UUID"),
+        // }
+        // TODO: #860 pull this from the DB
+        String::from("this is a placeholder lololol")
```
I can do this :)

yaleman replied:
I did it, pew pew removed.

Firstyear commented on kanidmd/score/src/https/middleware.rs:
```diff
@@ -0,0 +1,105 @@
+    /// set it in pages
+    pub fn new(domain_display_name: String) -> Self {
+        KanidmDisplayNameMiddleware {
+            // TODO: #860 work out how to get this out :D
```
Not sure about this middleware? Maybe for now, lets not put it in as a title, and do that in a second PR once we get the other domain name bits sorted?

yaleman replied:
Commented it out, since it's not actually used, implemented it slightly differently in the body - I hadn't gotten back to actually doing a full clean-up sweep when you looked at this >.>

Firstyear commented on kanidmd/score/src/https/mod.rs:
```diff
@@ -153,17 +157,22 @@ pub fn to_tide_response<T: Serialize>(
             <meta charset="utf-8"/>
             <meta name="viewport" content="width=device-width">
-            <title>Kanidm</title>
+            <title>===DOMAIN_DISPLAY_NAME===</title>
```
I think there has to be a better way than this, so maybe we leave this as is for now, and we talk about how a better method could look?

yaleman replied:
Tweaked it, this might be a little less terrible? I might be able to take the WASMHASH bit out of the middleware now I know how to mess around with  app state.

Firstyear commented on kanidmd/score/src/lib.rs:
```diff
@@ -390,6 +390,62 @@ pub fn vacuum_server_core(config: &Configuration) {
 }
 
+/// This'll set the domain's display name in the database
+pub fn set_domain_display_name(config: &Configuration, new_display_name: &str) {
```
Rather than doing this as an offline operation, why not just add domain_display_name as an access control for writes for admins, and then have a tool in kanidm_tools for it? My thinking is the "displayname" doesn't need offline edit, like you'll never need to change it in a DR to recover the situation (unlike domain name or admin password).

yaleman commented:
@Firstyear I can't work this out - I got it working where one could change it from the `kanidmd` CLI back in [602a25a](https://github.com/kanidm/kanidm/pull/872/commits/602a25a23bacf2682da1779b08d65263861810fb), but doing it through the API was blocked for anything I could try, I was getting `OperationError::SystemProtectedObject` which was coming from the `Protected` plugin... I think I've fixed that (adding to ALLOWED_ATTRS) in Protected but now I'm getting this:

```
ERROR    ┝━ 🚨 [admin.error]: Schema Violation in validation of modify_pre_apply InvalidAttribute("systemmay missing attribute: domain_display_name")
WARN     ┕━ ⚠️ [request.warn]: Client error --> Response sent | status: 400 - Bad Request
```

(I've updated the error messages while trying to find where it was coming from)

I can't figure out from the code or design what to do, so have to give up now, if you can fix it please go ahead 😭

Firstyear commented:
No problem @yaleman I'll check this out and will update/resolve :)

yaleman commented:
@Firstyear good to go now, rebased. I couldn't convince git to cherry-pick the other bits out without stomping things, hope that works OK. 

Thanks again for picking up my mess and fixing it 😻
